jdi-light/src/main/java/com/epam/jdi/light/elements/base/JdiSettings.java

```diff
@@ -1,22 +1,239 @@
+import static com.epam.jdi.light.settings.JDISettings.ELEMENT;
+import static com.epam.jdi.light.settings.WebSettings.*;
+import static com.epam.jdi.tools.EnumUtils.getEnumValue;
+import static com.epam.jdi.tools.LinqUtils.filter;
+import static com.epam.jdi.tools.ReflectionUtils.isClass;
+import static com.epam.jdi.tools.ReflectionUtils.isInterface;
+import static java.lang.String.format;
+import static java.util.Collections.singletonList;
+import static org.apache.commons.lang3.StringUtils.isBlank;
+
 public class JdiSettings {
-    public static JFunc2<JDIBase, Object[], WebElement> GET_WITH_ARGS = (b, args) -> {
+
+    public static JAction2<JDIBase, WebElement> VALIDATE_FOUND_ELEMENT = JdiSettings::validateFoundElement;
+
+    public static JFunc2<JDIBase, Object[], WebElement> GET_WITH_ARGS = JdiSettings::getWithArgs;
+    public static JFunc2<JDIBase, Object[], WebElement> GET_AND_VALIDATE = (b, args) -> {
+        WebElement element = GET_WITH_ARGS.execute(b, args);
+        logger.trace("Validate:" + element);
+        VALIDATE_FOUND_ELEMENT.execute(b, element);
+        b.beforeSearch(element);
+        return element;
+    };
+    private static WebElement getWithArgs(JDIBase b, Object[] args) {
+        logger.trace("getWithArgs");
         if (b.webElement.hasValue()) {
-            WebElement webElement = b.getCachedElement();
+            WebElement webElement = null;
+            try {
+                webElement = purify(b.webElement.get());
+                webElement.getTagName();
+            } catch (Exception ignore) { }
             if (webElement != null)
                 return webElement;
         }
-        if (b.locator.isNull())
-            return b.beforeSearch(b.getSmart());
+        if (b.locator.isNull() && b.parent == null)
+            return b.getSmart();
         if (b.locator.argsCount() != args.length)
-            return b.getWebListFromArgs(args);
-        List<WebElement> els = b.getAllElements(args);
-        return b.getElement(els);
-    };
+            return getUIElementFromArgs(b, args);
+        List<WebElement> els = getAllElementsInContext(b, args);
+        return filterWebListToWebElement(b, els);
+    }
+
+    // region Utilities
+    public static WebElement filterWebListToWebElement(JDIBase base, List<WebElement> els) {
+        if (els.size() == 1)
+            return els.get(0);
+        if (els.isEmpty())
+            throw exception(FAILED_TO_FIND_ELEMENT_MESSAGE, base.toString(), base.getTimeout());
+        List<WebElement> filtered = filterElements(base, els);
+        if (filtered.size() == 1)
+            return filtered.get(0);
+        if ((base.strictSearch == null && STRICT_SEARCH) || (base.strictSearch != null && base.strictSearch))
+            throw exception(FIND_TO_MUCH_ELEMENTS_MESSAGE, els.size(), base.toString(), base.getTimeout());
+        return (filtered.size() > 1 ? filtered : els).get(0);
+    }
+    public static final String FAILED_TO_FIND_ELEMENT_MESSAGE
```

Comment:
Codacy found an issue: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6397461)